src/Pim/Bundle/DataGridBundle/Extension/Sorter/Product/IsAssociatedSorter.php

```diff
@@ -14,11 +16,49 @@
  */
 class IsAssociatedSorter implements SorterInterface
 {
+    /** @var AssociationTypeRepositoryInterface */
+    protected $associationTypeRepository;
+
+    /** @var ProductRepositoryInterface */
+    protected $productRepository;
+
+    /**
+     * @param AssociationTypeRepositoryInterface $associationTypeRepository
+     * @param ProductRepositoryInterface         $productRepository
+     */
+    public function __construct(
+        AssociationTypeRepositoryInterface $associationTypeRepository,
+        ProductRepositoryInterface $productRepository
+    ) {
+        $this->associationTypeRepository = $associationTypeRepository;
+        $this->productRepository = $productRepository;
+    }
+
     /**
      * {@inheritdoc}
      */
     public function apply(DatasourceInterface $datasource, $field, $direction)
     {
-        $datasource->getProductQueryBuilder()->addSorter('is_associated', $direction);
+        $associationTypeId = $datasource->getParameters()['associationType'];
+        $currentProductId = $datasource->getParameters()['product'];
+
+        $associationType = null;
+        if (null !== $associationTypeId) {
+            $associationType = $this->associationTypeRepository->find($associationTypeId);
+        }
+
+        $product = null;
+        if (null !== $currentProductId) {
+            $product = $this->productRepository->find($currentProductId);
+        }
+
+        $datasource->getProductQueryBuilder()->addSorter(
+            sprintf(
+                'is_associated.%s.%s',
+                $associationType->getCode(),
```

Comment:
Dunno if a check is usefull as data is not put by the user but by the datagrid. If there is no association type, there is a big problem in the datagrid.
But I can throw a LogicException if you prefer.